modules/solutions/http-analytics/widgets/HTTPAnalyticsLatencyOverTime/src/utils/WidgetChannelManager.js

```diff
@@ -0,0 +1,120 @@
+        this.subscribeWidget = this.subscribeWidget.bind(this);
+        this.unsubscribeWidget = this.unsubscribeWidget.bind(this);
+        this._initializeWebSocket = this._initializeWebSocket.bind(this);
+        this._wsOnClose = this._wsOnClose.bind(this);
+        this._wsOnError = this._wsOnError.bind(this);
+        this._wsOnMessage = this._wsOnMessage.bind(this);
+        this._initializeWebSocket();
+        this.waitForConn = this.waitForConn.bind(this);
+    }
+
+    /**
+     * Set a widget to the widget map and send configuration to the provider endpoint.
+     * @param widgetId
+     * @param callback
+     * @param dataConfig
+     */
+    subscribeWidget(widgetId, callback, dataConfig) {
+        this.widgetMap[widgetId] = callback;
+        this.waitForConn(this.webSocket, () => {
+            let configJSON = {
+                providerName: dataConfig.type,
+                dataProviderConfiguration: dataConfig.config,
+                topic: widgetId,
+                action: 'subscribe',
+            };
+            this.webSocket.send(JSON.stringify(configJSON));
+        })
+    }
+
+    /**
+     * remove a widget from the widget map
+     * @param widgetId
+     */
+    unsubscribeWidget(widgetId) {
+        delete this.widgetMap[widgetId];
+        let config = {
+            topic: widgetId,
+            providerName: null,
+            dataProviderConfiguration: null,
+            action: 'unsubscribe',
+        };
+        this.webSocket.send(JSON.stringify(config));
+    }
+
+    /**
+     * Initialize websocket
+     * @private
+     */
+    _initializeWebSocket() {
+        this.webSocket = new WebSocket('wss://' + window.location.host + '/data-provider');
+        this.webSocket.onmessage = this._wsOnMessage;
+        this.webSocket.onerror = this._wsOnError;
+        this.webSocket.onclose = this._wsOnClose;
+    }
+
+    /**
+     * handle web-socket on message event
+     * @param message
+     * @private
+     */
+    _wsOnMessage(message) {
+        let data = JSON.parse(message.data);
+        if (this.widgetMap[data.topic]) {
+            this.widgetMap[data.topic](data);
+        } else {
+            console.log("Error in parsing message " + message);
+        }
+    }
+
+    /**
+     * handle web-socket on error event
+     * @param message
+     * @private
+     */
+    _wsOnError(message) {
+        console.log("Error in receiving message HTTPAnalyticsLatencyOverTime " + message);
+    }
+
+    /**
+     * handle web-socket on close event
+     * @param message
+     * @private
+     */
+    _wsOnClose(message) {
+        console.log("Close web socket event " + message);
+    }
+
+    waitForConn(socket, callback) {
+        let that = this;
+        setTimeout(() => {
+            if (socket.readyState === 1) {
+                if (callback !== null) {
+                    callback();
+                }
+            } else {
+                that.waitForConn(socket, callback);
```

Comment:
Check if it is possible to access `waitForConn` via `this` instead of assigning into `that`.

Reply:
Fixed in 737b4a5